server/src/ws/routes.ts

```diff
@@ -1,50 +1,117 @@
     async (connection, req) => {
       // A user join the chat
-      // const { id } = req.user as any
-      // const userAccounts = await usersStorageClient.load('allUsers')
-      // const user = userAccounts.find((account) => account.user.id === id)
-      // const username = user ? user.name : 'Unknown user'
-
-      broadcast({
-        sender: '__server',
-        message: `${'TODO'} joined`,
-      })
+      const { jwt } = req.query
+      const userId = getUserIdFromJwt(jwt)
+      const user = await getUserFromId(userId)
+      if (user) {
+        serverBroadcast(`${user.name} joined`)
+      }
 
       // A user leaving the chat
-      connection.socket.on('close', () => {
-        broadcast({
-          sender: '__server',
-          message: `${req.query.username} left`,
-        })
+      connection.socket.on('close', async () => {
+        const { jwt } = req.query
+        const userId = getUserIdFromJwt(jwt)
+        const user = await getUserFromId(userId)
+
+        if (user) {
+          serverBroadcast(`${user.name} left`)
+        }
       })
 
       // Broadcast incoming message
-      connection.socket.on('message', (message) => {
-        message = JSON.parse(message.toString())
+      connection.socket.on('message', async (msg) => {
+        msg = JSON.parse(msg.toString())
+        const incomingMessage = msg as unknown as {
+          threadId: string
+          message: string
+        }
+
+        const { jwt } = req.query
+        const userId: UserId = getUserIdFromJwt(jwt)
+
+        const res = await chatClient.AddMessageToThread(
+          incomingMessage.threadId,
+          userId,
+          incomingMessage.message
+        )
+
+        if (res.error) {
+          serverBroadcast(res.error)
+          return
+        }
+
+        broadcast(new WsMessageEvent('add', res.thread.id, res.message))
+
+        // If user is chatting with the bot:
+        if (
+          res.thread.participants.length === 2 &&
+          res.thread.participants.includes('autofriendid')
+        ) {
+          await sleep(3000)
+          const msg = chatGptClient
+            ? await chatGptClient.getResponse(incomingMessage.message)
+            : randChoice([
+                "That sound's great",
+                'Sure thing :)',
+                'How kind of you to say',
+              ])
+          const botRes = await chatClient.AddMessageToThread(
+            res.thread,
+            'autofriendid',
+            msg
+          )
+
+          if (res.error) {
+            serverBroadcast(res.error)
+            return
+          }
 
-        broadcast({
-          // sender: 'TODO',
-          ...message,
-        })
+          broadcast(new WsMessageEvent('add', botRes.thread.id, botRes.message))
+        }
       })
     }
   )
 
   // How do I let TS now that the server instance is already decorated with broadcast fn?
-  function broadcast(message) {
+  function broadcast(message: { [key: string]: any }) {
```

Comment:
You should try to use the type system that exists for this, and change this to message: WsEvent